File: src/main/java/com/redbox/domain/auth/filter/JWTFilter.java
package com.redbox.domain.auth.filter;

import com.redbox.domain.auth.dto.CustomUserDetails;
import com.redbox.domain.auth.exception.InvalidTokenException;
import com.redbox.domain.auth.util.JWTUtil;
import com.redbox.domain.user.entity.RoleType;
import com.redbox.domain.user.entity.User;
import com.redbox.global.exception.AuthException;
import com.redbox.global.exception.ErrorCode;
import com.redbox.global.util.error.ErrorResponseUtil;
import io.jsonwebtoken.ExpiredJwtException;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;

public class JWTFilter extends OncePerRequestFilter {

    private final JWTUtil jwtUtil;

    public JWTFilter(JWTUtil jwtUtil) {
        this.jwtUtil = jwtUtil;
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {

        // 헤더에서 access 키에 담긴 토큰을 가져옴
        String accessToken = request.getHeader("access");

        // 토큰이 없다면 다음 필터로 넘김
        if (accessToken == null) {
            filterChain.doFilter(request, response);
            return;
        }

        try {
            // 토큰 만료 여부 확인
            jwtUtil.isExpired(accessToken);

            // 토큰의 카테고리가 'access'인지 확인
            String category = jwtUtil.getCategory(accessToken);
            if (!"access".equals(category)) {
                throw new InvalidTokenException();
            }

            // email과 role 값을 토큰에서 가져옴
            Long userId = jwtUtil.getUserId(accessToken);
            String email = jwtUtil.getEmail(accessToken);
            String role = jwtUtil.getRole(accessToken);
            role = role.replace("ROLE_", "");

            User user = User.builder()
                    .id(userId)
                    .email(email)
                    .roleType(RoleType.valueOf(role)) // RoleType 설정
                    .build();

            CustomUserDetails customUserDetails = new CustomUserDetails(user);

            // 인증 정보 생성 후 SecurityContext에 저장
            Authentication authToken = new UsernamePasswordAuthenticationToken(
                    customUserDetails, null, customUserDetails.getAuthorities()
            );
            SecurityContextHolder.getContext().setAuthentication(authToken);

            // 다음 필터로 전달
            filterChain.doFilter(request, response);

        } catch (ExpiredJwtException e) {
            ErrorResponseUtil.handleException(response, ErrorCode.EXPIRED_TOKEN);
        } catch (AuthException e) {
            ErrorResponseUtil.handleException(response, e.getErrorCode());
        } catch (Exception e) {
            ErrorResponseUtil.handleException(response, ErrorCode.INTERNAL_SERVER_ERROR);
        }
    }
}

File: migration/src/main/kotlin/com/redbox/global/auth/filter/JWTFilter.kt
package com.redbox.global.auth.filter

import com.redbox.global.auth.util.JWTUtil
import com.redbox.domain.auth.dto.CustomUserDetails
import com.redbox.domain.user.user.repository.UserRepository
import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.MalformedJwtException
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import java.security.SignatureException

@Component
class JWTFilter(
    private val jwtUtil: JWTUtil,
    private val userRepository: UserRepository
) : OncePerRequestFilter() {

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        val token = getTokenFromRequest(request)

        if (token != null) {
            try {
                if (!jwtUtil.isExpired(token)) {
                    val email = jwtUtil.getEmail(token)
                    val user = userRepository.findByEmail(email)

                    user?.let {
                        val userDetails: UserDetails = CustomUserDetails(it)
                        val authentication = UsernamePasswordAuthenticationToken(
                            userDetails, null, userDetails.authorities
                        )
                        SecurityContextHolder.getContext().authentication = authentication
                    }
                }
            } catch (e: ExpiredJwtException) {
                response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "JWT 토큰이 만료되었습니다.")
                return
            } catch (e: SignatureException) {
                response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "JWT 서명이 유효하지 않습니다.")
                return
            } catch (e: MalformedJwtException) {
                response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "잘못된 JWT 형식입니다.")
                return
            }
        }

        filterChain.doFilter(request, response)
    }

    private fun getTokenFromRequest(request: HttpServletRequest): String? {
        val bearerToken = request.getHeader("Authorization")
        return if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
            bearerToken.substring(7)
        } else null
    }
}
